### DISCUSSION TITLE
How to change themes(not styles like dark/light) dynamically ?

### DISCUSSION THREAD

**Discussion by AkiraRy** [TomSchimansky/CustomTkinter#1824] [Q&A] [answered]
`import customtkinter

customtkinter.set_appearance_mode("dark")  # Modes: "System" (standard), "Dark", "Light"
customtkinter.set_default_color_theme("blue")  # Themes: "blue" (standard), "green", "dark-blue"

app = customtkinter.CTk()
app.geometry("400x780")
app.title("CustomTkinter simple_example.py")

def button_callback():
    print("Button click", combobox_1.get())
    customtkinter.set_default_color_theme("pinky.json")  # Themes: "blue" (standard), "green", "dark-blue"

def slider_callback(value):
    progressbar_1.set(value)

frame_1 = customtkinter.CTkFrame(master=app)
frame_1.pack(pady=20, padx=60, fill="both", expand=True)

label_1 = customtkinter.CTkLabel(master=frame_1, justify=customtkinter.LEFT)
label_1.pack(pady=10, padx=10)

progressbar_1 = customtkinter.CTkProgressBar(master=frame_1)
progressbar_1.pack(pady=10, padx=10)

button_1 = customtkinter.CTkButton(master=frame_1, command=button_callback)
button_1.pack(pady=10, padx=10)

slider_1 = customtkinter.CTkSlider(master=frame_1, command=slider_callback, from_=0, to=1)
slider_1.pack(pady=10, padx=10)
slider_1.set(0.5)

entry_1 = customtkinter.CTkEntry(master=frame_1, placeholder_text="CTkEntry")
entry_1.pack(pady=10, padx=10)

optionmenu_1 = customtkinter.CTkOptionMenu(frame_1, values=["Option 1", "Option 2", "Option 42 long long long..."])
optionmenu_1.pack(pady=10, padx=10)
optionmenu_1.set("CTkOptionMenu")

combobox_1 = customtkinter.CTkComboBox(frame_1, values=["Option 1", "Option 2", "Option 42 long long long..."])
combobox_1.pack(pady=10, padx=10)
combobox_1.set("CTkComboBox")

checkbox_1 = customtkinter.CTkCheckBox(master=frame_1)
checkbox_1.pack(pady=10, padx=10)

radiobutton_var = customtkinter.IntVar(value=1)

radiobutton_1 = customtkinter.CTkRadioButton(master=frame_1, variable=radiobutton_var, value=1)
radiobutton_1.pack(pady=10, padx=10)

radiobutton_2 = customtkinter.CTkRadioButton(master=frame_1, variable=radiobutton_var, value=2)
radiobutton_2.pack(pady=10, padx=10)

switch_1 = customtkinter.CTkSwitch(master=frame_1)
switch_1.pack(pady=10, padx=10)

text_1 = customtkinter.CTkTextbox(master=frame_1, width=200, height=70)
text_1.pack(pady=10, padx=10)
text_1.insert("0.0", "CTkTextbox\n\n\n\n")

segmented_button_1 = customtkinter.CTkSegmentedButton(master=frame_1, values=["CTkSegmentedButton", "Value 2"])
segmented_button_1.pack(pady=10, padx=10)

tabview_1 = customtkinter.CTkTabview(master=frame_1, width=200, height=70)
tabview_1.pack(pady=10, padx=10)
tabview_1.add("CTkTabview")
tabview_1.add("Tab 2")

app.mainloop()`  
this is example code, how do i make it to change the theme using my custom json files? When i press the button it doesn't change the theme and yes if i change the theme manually before the gui starts it will change, thanks

**Comment by avalon60** [comment, marked as answer]
It's not a straight forward task. You would need to teardown and rebuild the widgets, to get the new theme to take effect.

One way would be to maintain a list of all the widgets you create, and use that to cycle through and destroy the widgets, and have a dedicated method/function to recreate them. This may present its own challenges, if you don't destroy them in the correct order, although I might be wrong on this particular point. For example if you attempt to destroy a button, after you have already destroyed its parent widget. You could of course catch and ignore the raised exceptions. 

Another consideration is that you can't destroy the CTk() master widget. CTk() has an fg_color property, so you would need to do a look-up (there's a CTkThemeManager api which facilitates this.) and grab the colour code from the new theme, and explicitly configure the CTk() widget, to update it to the new colour.

Finally, much depends on your application design. Depending on what your app is doing, you may need to store the session state (data changes etc. made by the user). You would then need to restore these, after rebuilding the destroyed widgets. You might be lucky, and have an app, where CTk() itself, doesn't directly host any data widgets.

**Comment by AkiraRy** [comment]
It is a lot of work , i will better give the user choice at the beginning before starting the gui. thanks for your answer

**Comment by AkiraRy** [comment]
It is not connected to the main topic but i have question regarding grid and frame in ctk library , here is the example of my code 
`import customtkinter
import os
from PIL import Image
customtkinter.set_default_color_theme("theme.json")

class App(customtkinter.CTk):
    width = 900
    height = 600

    def __init__(self):
        super().__init__()
        self.default_host = False
        self.default_whisper = False
        self.iconbitmap('amadeus.ico')
        self.title('Kurisu control panel')
        self.geometry(f"{self.width}x{self.height}")

        #bg_image
        current_path = os.path.dirname(os.path.realpath(__file__))
        self.bg_image = customtkinter.CTkImage(Image.open(current_path + "/test_images/black_bg.webp"),
                                               size=(self.width, self.height))
        self.bg_image_label = customtkinter.CTkLabel(self, text='', image=self.bg_image)
        self.bg_image_label.grid(row=0, column=0)

        # create login frame
        self.login_frame = customtkinter.CTkFrame(self, corner_radius=0)
        self.login_frame.grid(row=0, column=0, sticky="ns")

        self.login_label = customtkinter.CTkLabel(self.login_frame, text="Starting page",
                                                  font=customtkinter.CTkFont(size=20, weight="bold"))
        self.login_label.grid(row=0, column=0, padx=30, pady=(150, 15), columnspan=2)
        self.host_entry = customtkinter.CTkEntry(self.login_frame, width=200, placeholder_text='host')
        self.host_entry.grid(row=1, column=0, padx=30, pady=(15, 15))
        self.whisper_host = customtkinter.CTkEntry(self.login_frame, width=200, placeholder_text="whisper host")
        self.whisper_host.grid(row=2, column=0, padx=30, pady=(0, 15))
        self.launch_button = customtkinter.CTkButton(self.login_frame, text="Launch", command=self.launch_event, width=200)
        self.launch_button.grid(row=3, column=0, padx=30, pady=(15, 15))

        self.checkbox_1 = customtkinter.CTkCheckBox(master=self.login_frame, text='')
        self.checkbox_1.grid(row=1, column=2,  pady=(9, 15))


        # create main frame
        self.main_frame = customtkinter.CTkFrame(self, corner_radius=0,  width=200, height=10)
        self.main_frame.grid_columnconfigure(0, weight=1)


    def launch_event(self):


        print("Launch pressed\nHost:", self.host_entry.get(), "\nWhisperH ost:", self.whisper_host.get())

        self.login_frame.grid_forget()  # remove login frame
        self.main_frame.grid(row=0, column=0, sticky="nsew", padx=100)  # show main frame

    def back_event(self):
        self.main_frame.grid_forget()  # remove main frame
        self.login_frame.grid(row=0, column=0, sticky="ns")  # show login frame

if __name__ == "__main__":
    app = App()
    app.mainloop()` 
    
![image](https://github.com/TomSchimansky/CustomTkinter/assets/106382882/298e7a8e-8729-4f9a-82b4-77f5ed7c84d8)
how can i decrease the gap between checkbox and right side of frame? I`ve tried a lot of things like padx,x ipadx, sticky, columnspan, even changing the entry and checkbox places or changing the pading of mail lable Starting, but it still does have this gap and i have no idea why

**Reply by avalon60** [reply]
I'd need to play with the code, but out of interest, what happens if you replace the checkbox with tk.Checkbutton?

**Reply by AkiraRy** [reply]
![image](https://github.com/TomSchimansky/CustomTkinter/assets/106382882/199ce4f8-b454-4fd3-91a9-56a9ae5dbf20)
now it looks better, but it is still kinda uneven the gap between entry-checkbox and checkbox-right_side_of_frame
i`ve tried playing with padx/ipadx but it still persist

**Reply by AkiraRy** [reply]
my thery is that CTkCheckbox has default width reserverd for text so this is why there is a gap but changing the width or text value of this checkbox doesn`t seems to aeffect anything

**Reply by AkiraRy** [reply]
nevermind I didn`t changed the width of checkbox from 100 to prefered, who thought this width also included the space for text

**Reply by avalon60** [reply]
Well, the reason I suggested you try that was because I was wondering the same, ie it was CTKCheckBox causing the problem, because it renders a label. In your case the label is empty, but it may nevertheless be taking some space. You could try raising an issue.

**Comment by AkiraRy** [comment]
the main purpose of the program is to get 2 variables from the user (it will be endpoings from service like colab), after that i start the application with button(it was in the example but i`ve deleted it for time being) and it swithces to another frame where i can operate my program
